### PR TITLE
[RzIL] Refactor RzAnalysis side of IL

### PR DESCRIPTION
**Pull request by thestr4ng3r**
## Do not squash

**Your checklist for this pull request**
- [x] I've read the [guidelines for contributing](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md) to this repository
- [x] I made sure to follow the project's [coding style](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md#code-style)
- [x] I've documented or updated the documentation of every function and struct this PR changes. If not so I've explained why.
- [x] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the [rizin book](https://github.com/rizinorg/book) with the relevant information (if needed)

**Detailed description**

This refactors primarily the analysis side of RzIL to achieve the following design:
* `RzILVM` is a low-level IL vm. It does not have any connection to real architectures, registers, etc. by itself
* `RzAnalysisILVM` is a self-contained vm that internally uses `RzILVM` to emulate a real architecture using `RzAnalysis` to disassemble and lift. It also manages the binding to registers and syncing.

In addition, analysis plugins now don't actively initialize the vm anymore, but return a fully declarative description `RzAnalysisILConfig` of how to set up the vm and optionally also its initial state as `RzAnalysisILInitState`.

This also enables multiple IL vms to exist at the same time as plugins can not mess with the global vm anymore and local `RzReg` instances can be used. As a sidenote, one thing that can't be localized easily right now however is io. It is theoretically possible to use a sparse overlay buffer over the io buffer to make all writes local, but then it is not possible to use the finer-grained per-map permissions of `RzIO`, so depends on the use-case if that is sufficient.

**Test plan**

See the added integration test for an example of using an independent `RzAnalysisILVM`.

### CODE REVIEW

File: librz/analysis/il/analysis_il.c [hunk 1]
```diff
@@ -0,0 +1,276 @@
+// SPDX-FileCopyrightText: 2021 heersin <teablearcher@gmail.com>
+// SPDX-License-Identifier: LGPL-3.0-only
+
+#include <rz_analysis.h>
+
+/////////////////////////////////////////////////////////
```
Comment (wargio) [inline librz/analysis/il/analysis_il.c]: are these necessary?

File: librz/il/definitions/label.c [hunk 2]
```diff
@@ -18,3 +12,29 @@ RZ_API RzILEffectLabel *rz_il_effect_label_new(RZ_NONNULL const char *name, RzIL
 	lbl->type = type;
 	return lbl;
 }
+
+RZ_API void rz_il_effect_label_free(RzILEffectLabel *lbl) {
```
Comment (XVilka) [inline librz/il/definitions/label.c:16]: Missing Doxygen

File: librz/analysis/il/analysis_il.c (hunk 1 above)
Comment (thestr4ng3r) [reply librz/analysis/il/analysis_il.c]: As visual separators for me, yes. Not this first one though.

File: librz/analysis/il/analysis_il.c [hunk 3]
```diff
@@ -0,0 +1,275 @@
...
+
+static void var_state_free(void *e, void *user) {
+	RzAnalysisILInitStateVar *s = e;
+	if (!s) {
+		return;
+	}
+	rz_il_value_free(s->val);
+}
+
+RZ_API RzAnalysisILInitState *rz_analysis_il_init_state_new() {
```
Comment (XVilka) [inline librz/analysis/il/analysis_il.c:19]: Doxygen, please

File: librz/analysis/il/analysis_il.c [hunk 4]
```diff
@@ -0,0 +1,275 @@
...
+}
+
+RZ_API void rz_analysis_il_init_state_free(RzAnalysisILInitState *state) {
+	if (!state) {
+		return;
+	}
+	rz_vector_fini(&state->vars);
+}
+
+RZ_API void rz_analysis_il_init_state_set_var(RZ_NONNULL RzAnalysisILInitState *state,
```
Comment (XVilka) [inline librz/analysis/il/analysis_il.c:38]: Especially for this one

File: librz/analysis/il/analysis_il.c [hunk 5]
```diff
@@ -0,0 +1,275 @@
...
+	}
+	if (is) {
+		RzAnalysisILInitStateVar *v;
+		rz_vector_foreach(&is->vars, v) {
+			rz_il_vm_set_global_var(vm->vm, v->name, rz_il_value_dup(v->val));
+		}
+	}
+}
+
+RZ_API void rz_analysis_il_vm_sync_from_reg(RzAnalysisILVM *vm, RZ_NONNULL RzReg *reg) {
```
Comment (XVilka) [inline librz/analysis/il/analysis_il.c:208]: All those API functions are important enough to have doxygne.

File: librz/core/cil.c [hunk 6]
```diff
@@ -510,8 +509,8 @@ static void rzil_print_register_bitv(RzBitVector *number, ILPrint *p) {
 }
 
 RZ_IPI void rz_core_analysis_rzil_vm_status(RzCore *core, const char *var_name, RzOutputMode mode) {
```
Comment (XVilka) [inline librz/core/cil.c]: then probably rename this function too

File: librz/core/cil.c [hunk 7]
```diff
@@ -609,58 +608,40 @@ RZ_IPI void rz_core_analysis_rzil_vm_status(RzCore *core, const char *var_name,
  * \return false if an error occured (e.g. invalid op)
  */
 RZ_IPI bool rz_core_rzil_step(RzCore *core) {
```
Comment (XVilka) [inline librz/core/cil.c]: And all those `rz_core_analysis_rzil_*` functions, to be consistent with naming changes you did.

File: librz/core/cil.c [hunk 8]
```diff
@@ -609,58 +608,40 @@ RZ_IPI void rz_core_analysis_rzil_vm_status(RzCore *core, const char *var_name,
...
+	return false;
 }
 
-RZ_IPI void rz_core_analysis_rzil_step_with_events(RzCore *core, PJ *pj) {
-	rz_core_rzil_step(core);
+/**
+ * Perform a single step at the PC given by analysis->reg in RzIL and print any events that happened
+ * \return false if an error occured (e.g. invalid op)
+ */
+RZ_IPI bool rz_core_analysis_rzil_step_with_events(RzCore *core, PJ *pj) {
```
Comment (XVilka) [inline librz/core/cil.c]: and this one

**Review by XVilka — CHANGES_REQUESTED** [review]

**Comment by XVilka** [comment]
Good idea to separate those two concepts 👍 
In fact, may be worth having a paragraph about this in `doc/rzil.md`.

File: test/integration/test_analysis_il.c [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+// SPDX-License-Identifier: LGPL-3.0-only
+
+#include <rz_core.h>
+#include "../unit/minunit.h"
+
+/**
+ * Test running an IL vm with Analysis connection independently of the global user-faced vm
+ */
+static bool test_analysis_il_vm_step() {
+	RzCore *core = rz_core_new();
```
Comment (ret2libc) [inline test/integration/test_analysis_il.c:11]: dumb question, sorry: is it possible to use the IL VM without Core?

Comment (thestr4ng3r) [reply test/integration/test_analysis_il.c:11]: Yes, should be possible. You will currently have to set up io manually though to allow it to disassemble.

File: librz/il/definitions/label.c (hunk 2 above)
Comment (thestr4ng3r) [reply librz/il/definitions/label.c:16]: This is a free function. What kind of docs do you expect for this that doesn't just repeat what the name already says?

File: librz/analysis/il/analysis_il.c (hunk 3 above)
Comment (thestr4ng3r) [reply librz/analysis/il/analysis_il.c:19]: same

File: librz/il/definitions/label.c (hunk 2 above)
Comment (ret2libc) [reply librz/il/definitions/label.c:16]: Agree, kinda useless in these simple cases.

File: test/integration/test_analysis_il.c (hunk 9 above)
Comment (ret2libc) [reply test/integration/test_analysis_il.c:11]: Ok thanks

**Review by XVilka — APPROVED** [review]
Good to go!

**Pull request merged**